### PR TITLE
Add a backend based on C++11 std::thread in place of POSIX threads

### PR DESCRIPTION
**Pull request by PhilMiller**
Replaces all POSIX-specific types and function calls with notionally portable C++ standard library equivalents. This should enable broader performance-portable use on Windows and possibly some other non-*nix-like platforms.

- `pthread_t` -> `std::thread` and `std::thread::id`
- `pthread_attr_setdetachable()` -> `thread.detach()`
- `pthread_self()` -> `std::this_thread::get_id()`
- `pthread_mutex_t` -> `std::mutex` (with associated `lock()` and `unlock()` calls)
- `sched_yield()` -> `std::this_thread::yield()`

### CODE REVIEW

**Comment by PhilMiller** [comment]
Now that all the tests pass, this should be ready for performance benchmarking. I'll make one more commit to swap out the two implementations, that can easily be used or reverted, to make comparison as trivial as possible.

**Comment by ndellingwood** [comment]
I tried testing this PR standalone and within Trilinos using gcc/7.2.0, I ran into compilation errors in both cases as follows:

```
/ascldap/users/ndellin/kokkos/core/src/CppThreads/Kokkos_CppThreadsExec_base.cpp:72:6: error: ‘mutex’ in namespace ‘std’ does not name a type
 std::mutex host_internal_cppthread_mutex;
      ^~~~~
/ascldap/users/ndellin/kokkos/core/src/CppThreads/Kokkos_CppThreadsExec_base.cpp: In static member function ‘static void Kokkos::Impl::ThreadsExec::global_lock()’:
/ascldap/users/ndellin/kokkos/core/src/CppThreads/Kokkos_CppThreadsExec_base.cpp:114:35: error: ‘host_internal_cppthread_mutex’ was not declared in this scope
 void ThreadsExec::global_lock() { host_internal_cppthread_mutex.lock(); }
                                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/ascldap/users/ndellin/kokkos/core/src/CppThreads/Kokkos_CppThreadsExec_base.cpp: In static member function ‘static void Kokkos::Impl::ThreadsExec::global_unlock()’:
/ascldap/users/ndellin/kokkos/core/src/CppThreads/Kokkos_CppThreadsExec_base.cpp:116:37: error: ‘host_internal_cppthread_mutex’ was not declared in this scope
 void ThreadsExec::global_unlock() { host_internal_cppthread_mutex.unlock(); }
                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
make[2]: *** [core/src/CMakeFiles/kokkoscore.dir/CppThreads/Kokkos_CppThreadsExec_base.cpp.o] Error 1
```

Here was my Kokkos configure:

```
cmake -DCMAKE_CXX_COMPILER=g++ -DCMAKE_INSTALL_PREFIX=$KOKKOS_INSTALL -DKokkos_ENABLE_PTHREAD=ON -DKokkos_ARCH_BDW=ON -DKokkos_ENABLE_TESTS=ON -DKokkos_ENABLE_EXAMPLES=OFF -DCMAKE_VERBOSE_MAKEFILE=ON -DCMAKE_CXX_EXTENSIONS=OFF -DCMAKE_BUILD_TYPE=RELEASE $KOKKOS_PATH
```

@crtrott regarding Trilinos testing, I don't think there is any regular testing of Pthreads for pull requests, and definitely not for the configurations using the ATDM script framework (setting pthread as option errors out). I hacked the scripts in my local Trilinos repo to enable Threads testing, but wanted to let you know currently there seems to be little active testing

**Review by crtrott — CHANGES_REQUESTED** [review]
We can discuss this tomorrow in the meeting, but I really was thinking about just changing the posix threads backend, and not having it anymore. I think there is no practical good reason to keep it around, and good reasons not to. As an intermediate step this is good, but we don't need the old implementation.

**Comment by PhilMiller** [comment]
Per @crtrott check that this still works with HWLOC enabled

File: core/src/CppThreads/Kokkos_CppThreadsExec.cpp [hunk 1]
```diff
@@ -0,0 +1,855 @@
...
+
+#include <Kokkos_Macros.hpp>
+#if defined(KOKKOS_ENABLE_THREADS)
+
+#include <cstdint>
+#include <limits>
+#include <utility>
+#include <iostream>
+#include <sstream>
+#include <thread>
```
Comment (PhilMiller) [inline core/src/CppThreads/Kokkos_CppThreadsExec.cpp:53]: missing `#include <mutex>`

**Comment by PhilMiller** [comment]
test this please

**Comment by PhilMiller** [comment]
retest this please

**Comment by dalg24** [comment]
Retest this please

**Comment by PhilMiller** [comment]
examinimd-threads-comparison
Here's a rough performance comparison between old and new on kokkos-dev-2. If there are any differences, they're tiny, and favor the new `std::thread` implementation

**Comment by PhilMiller** [comment]
Key question for @crtrott or whoever - do we want to rename and/or alias the options used to choose or describe this backend in the build system and code base? Many things are still anachronistically labeled 'pthread' with this code change

File: BUILD.md [hunk 2]
```diff
@@ -120,7 +120,7 @@ Device backends can be enabled by specifying `-DKokkos_ENABLE_X`.
     * Whether to build OpenMP backend
     * BOOL Default: OFF
 * Kokkos_ENABLE_PTHREAD
```
Comment (PhilMiller) [inline BUILD.md:122]: The name here is an anachronism

**Comment by masterleinad** [comment]
> Key question for @crtrott or whoever - do we want to rename and/or alias the options used to choose or describe this backend in the build system and code base? Many things are still anachronistically labeled 'pthread' with this code change

Deprecating `pthreads` in favor of `threads` came up recently and I think this is the appropriate place to do that.

**Comment by PhilMiller** [comment]
> > Key question for crtrott or whoever - do we want to rename and/or alias the options used to choose or describe this backend in the build system and code base? Many things are still anachronistically labeled 'pthread' with this code change
> 
> Deprecating `pthreads` in favor of `threads` came up recently and I think this is the appropriate place to do that.

@masterleinad 

If that's the course we'd like to take, I'm happy to do the work, but I'd suggest a follow-on PR for clarity of review, to keep this one focused on the functional/substantive changes involved

**Comment by crtrott** [comment]
Agree with you Phil.

**Comment by PhilMiller** [comment]
Testing HWLOC enablement on Blake:

```
$ module list
Currently Loaded Modulefiles:
  1) git/2.9.4         2) binutils/2.30.0   3) gcc/9.2.0         4) numa/2.0.11       5) hwloc/1.11.8      6) ninja/1.7.2       7) cmake/3.19.3

$ cmake ~/repos/kokkos/  -G Ninja  -DCMAKE_BUILD_TYPE=Release  -DKokkos_ENABLE_PTHREAD=ON -DKokkos_ENABLE_HWLOC=ON -DKokkos_ENABLE_TESTS=ON -DKokkos_ENABLE_EXAMPLES=ON
...
$ ninja
...
$ ctest
...
100% tests passed, 0 tests failed out of 41
```

**Comment by crtrott** [comment]
Windows with two tiny changes discussed on slack to not enable  libpthread dependency and turn that one test off:

```
Test project C:/Users/ceear/source/repos/kokkos/out/build/x64-Clang-Pthreads
      Start  1: KokkosCore_UnitTest_Threads
 1/36 Test  #1: KokkosCore_UnitTest_Threads .................   Passed   15.98 sec
      Start  2: KokkosCore_UnitTest_Default
 2/36 Test  #2: KokkosCore_UnitTest_Default .................   Passed    0.07 sec
      Start  3: KokkosCore_UnitTest_PushFinalizeHook
 3/36 Test  #3: KokkosCore_UnitTest_PushFinalizeHook ........   Passed    0.05 sec
      Start  4: KokkosCore_UnitTest_Develop
 4/36 Test  #4: KokkosCore_UnitTest_Develop .................   Passed    0.04 sec
      Start  5: KokkosCore_UnitTest_LogicalSpaces
 5/36 Test  #5: KokkosCore_UnitTest_LogicalSpaces ...........   Passed    0.07 sec
      Start  6: KokkosCore_UnitTest_EventCorrectness
 6/36 Test  #6: KokkosCore_UnitTest_EventCorrectness ........   Passed    0.04 sec
      Start  7: KokkosCore_UnitTest_StackTraceTest
 7/36 Test  #7: KokkosCore_UnitTest_StackTraceTest ..........   Passed    0.09 sec
      Start  8: KokkosCore_UnitTest_DefaultInit_1
 8/36 Test  #8: KokkosCore_UnitTest_DefaultInit_1 ...........   Passed    0.03 sec
      Start  9: KokkosCore_UnitTest_DefaultInit_2
 9/36 Test  #9: KokkosCore_UnitTest_DefaultInit_2 ...........   Passed    0.04 sec
      Start 10: KokkosCore_UnitTest_DefaultInit_3
10/36 Test #10: KokkosCore_UnitTest_DefaultInit_3 ...........   Passed    0.04 sec
      Start 11: KokkosCore_UnitTest_DefaultInit_4
11/36 Test #11: KokkosCore_UnitTest_DefaultInit_4 ...........   Passed    0.03 sec
      Start 12: KokkosCore_UnitTest_DefaultInit_5
12/36 Test #12: KokkosCore_UnitTest_DefaultInit_5 ...........   Passed    0.03 sec
      Start 13: KokkosCore_UnitTest_DefaultInit_6
13/36 Test #13: KokkosCore_UnitTest_DefaultInit_6 ...........   Passed    0.04 sec
      Start 14: KokkosCore_UnitTest_DefaultInit_7
14/36 Test #14: KokkosCore_UnitTest_DefaultInit_7 ...........   Passed    0.03 sec
      Start 15: KokkosCore_UnitTest_DefaultInit_8
15/36 Test #15: KokkosCore_UnitTest_DefaultInit_8 ...........   Passed    0.04 sec
      Start 16: KokkosCore_UnitTest_DefaultInit_9
16/36 Test #16: KokkosCore_UnitTest_DefaultInit_9 ...........   Passed    0.03 sec
      Start 17: KokkosCore_UnitTest_DefaultInit_10
17/36 Test #17: KokkosCore_UnitTest_DefaultInit_10 ..........   Passed    0.04 sec
      Start 18: KokkosCore_UnitTest_DefaultInit_11
18/36 Test #18: KokkosCore_UnitTest_DefaultInit_11 ..........   Passed    0.03 sec
      Start 19: KokkosCore_UnitTest_DefaultInit_12
19/36 Test #19: KokkosCore_UnitTest_DefaultInit_12 ..........   Passed    0.03 sec
      Start 20: KokkosCore_UnitTest_DefaultInit_13
20/36 Test #20: KokkosCore_UnitTest_DefaultInit_13 ..........   Passed    0.04 sec
      Start 21: KokkosCore_UnitTest_DefaultInit_14
21/36 Test #21: KokkosCore_UnitTest_DefaultInit_14 ..........   Passed    0.03 sec
      Start 22: KokkosCore_UnitTest_DefaultInit_15
22/36 Test #22: KokkosCore_UnitTest_DefaultInit_15 ..........   Passed    0.03 sec
      Start 23: KokkosCore_UnitTest_DefaultInit_16
23/36 Test #23: KokkosCore_UnitTest_DefaultInit_16 ..........   Passed    0.04 sec
      Start 24: KokkosCore_UnitTest_DefaultInit_17
24/36 Test #24: KokkosCore_UnitTest_DefaultInit_17 ..........   Passed    0.03 sec
      Start 25: KokkosCore_UnitTest_DefaultInit_18
25/36 Test #25: KokkosCore_UnitTest_DefaultInit_18 ..........   Passed    0.03 sec
      Start 26: KokkosCore_IncrementalTest_PTHREAD
26/36 Test #26: KokkosCore_IncrementalTest_PTHREAD ..........   Passed    0.03 sec
      Start 27: KokkosCore_UnitTest_CTestDevice
27/36 Test #27: KokkosCore_UnitTest_CTestDevice .............   Passed    0.04 sec
      Start 28: KokkosCore_UnitTest_CMakePassCmdLineArgs0
28/36 Test #28: KokkosCore_UnitTest_CMakePassCmdLineArgs0 ...   Passed    0.03 sec
      Start 29: KokkosCore_PerfTestExec
29/36 Test #29: KokkosCore_PerfTestExec .....................   Passed  115.16 sec
      Start 30: KokkosCore_PerformanceTest_Atomic
30/36 Test #30: KokkosCore_PerformanceTest_Atomic ...........   Passed    0.09 sec
      Start 31: KokkosCore_PerformanceTest_Atomic_MinMax
31/36 Test #31: KokkosCore_PerformanceTest_Atomic_MinMax ....   Passed    0.36 sec
      Start 32: KokkosCore_PerformanceTest_Mempool
32/36 Test #32: KokkosCore_PerformanceTest_Mempool ..........   Passed    0.03 sec
      Start 33: KokkosCore_PerformanceTest_TaskDag
33/36 Test #33: KokkosCore_PerformanceTest_TaskDag ..........   Passed    0.03 sec
      Start 34: KokkosContainers_UnitTest_Threads
34/36 Test #34: KokkosContainers_UnitTest_Threads ...........   Passed  376.16 sec
      Start 35: KokkosContainers_PerformanceTest_Threads
35/36 Test #35: KokkosContainers_PerformanceTest_Threads ....   Passed   65.47 sec
      Start 36: KokkosAlgorithms_UnitTest
36/36 Test #36: KokkosAlgorithms_UnitTest ...................   Passed   15.46 sec
100% tests passed, 0 tests failed out of 36
Total Test time (real) = 589.91 sec
```

**Review by crtrott — APPROVED** [review]

**Comment by PhilMiller** [comment]
10k steps of ExaMiniMD on stria (ARM aarch64) and no crash. I think we can remove the remark in the README about ARM/Threads being known non-working

**Comment by PhilMiller** [comment]
> 10k steps of ExaMiniMD on stria (ARM aarch64) and no crash. I think we can remove the remark in the README about ARM/Threads being known non-working

This will be deferred until #4322 is addressed, to be more confident in the results.

File: core/src/Threads/Kokkos_ThreadsExec_base.cpp [hunk 3]
```diff
@@ -96,140 +92,29 @@ void* internal_pthread_driver(void*) {
...
-
-    result = 0 == pthread_create(&pt, &attr, internal_pthread_driver, nullptr);
-  }
+  std::thread t(internal_cppthread_driver);
+  t.detach();
 
-  pthread_attr_destroy(&attr);
-
-  return result;
+  return true;
```
Comment (dalg24) [inline core/src/Threads/Kokkos_ThreadsExec_base.cpp:98]: Should we still be returning anything from the function then?

File: core/src/CMakeLists.txt [hunk 4]
```diff
@@ -95,7 +95,10 @@ KOKKOS_LINK_TPL(kokkoscore PUBLIC CUDA)
 KOKKOS_LINK_TPL(kokkoscore PUBLIC HPX)
 KOKKOS_LINK_TPL(kokkoscore PUBLIC LIBDL)
 KOKKOS_LINK_TPL(kokkoscore PUBLIC LIBRT)
-KOKKOS_LINK_TPL(kokkoscore PUBLIC PTHREAD)
+# On *nix-like systems (Linux, macOS) we need pthread for C++ std::thread
+IF (NOT WIN32)
+  KOKKOS_LINK_TPL(kokkoscore PUBLIC PTHREAD)
+ENDIF()
```
[the remark this answers is not in the document]
Comment (PhilMiller) [reply core/src/CMakeLists.txt:101]: That's what Christian and I thought, too, but tests failed to link without it on Linux

File: core/src/Threads/Kokkos_ThreadsExec_base.cpp (hunk 3 above)
Comment (PhilMiller) [reply core/src/Threads/Kokkos_ThreadsExec_base.cpp:98]: I guess not. I was being minimalist in my changes, but I could simplify that

File: core/unit_test/TestPolicyConstruction.hpp [hunk 5]
```diff
@@ -790,6 +790,7 @@ struct static_assert_dummy_policy_must_be_size_of_desired_occupancy<
     sizeof(Kokkos::Experimental::DesiredOccupancy),
     sizeof(Kokkos::Experimental::DesiredOccupancy)> {};
 
+#ifndef _WIN32
```
[the remark this answers is not in the document]
Comment (PhilMiller) [reply core/unit_test/TestPolicyConstruction.hpp:793]: Per @crtrott I think so. This was at his direction.

[the remark this answers is not in the document]
Comment (dalg24) [reply core/unit_test/TestPolicyConstruction.hpp:793]: In any case I don't think it belongs to this PR

File: core/src/Threads/Kokkos_ThreadsExec.cpp [hunk 6]
```diff
@@ -870,7 +873,7 @@ void ThreadsSpaceInitializer::print_configuration(std::ostream &msg,
                                                   const bool detail) {
   msg << "Host Parallel Execution Space:" << std::endl;
   msg << "  KOKKOS_ENABLE_THREADS: ";
-  msg << "yes" << std::endl;
+  msg << "yes (C++ std::thread)" << std::endl;
```
[the remark this answers is not in the document]
Comment (PhilMiller) [reply core/src/Threads/Kokkos_ThreadsExec.cpp]: Undone

File: README.md [hunk 7]
```diff
@@ -107,7 +107,7 @@ For specifics see the LICENSE file contained in the repository or distribution.
 
 Primary tested compiler are passing in release mode
 with warnings as errors. They also are tested with a comprehensive set of
-backend combinations (i.e. OpenMP, Pthreads, Serial, OpenMP+Serial, ...).
+backend combinations (i.e. OpenMP, threads, Serial, OpenMP+Serial, ...).
```
[the remark this answers is not in the document]
Comment (PhilMiller) [reply README.md]: Done

File: core/unit_test/TestPolicyConstruction.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (PhilMiller) [reply core/unit_test/TestPolicyConstruction.hpp:793]: It's necessary to achieve the goal of the Threads backend being usable on Windows, part of the motivation for this work

File: core/src/CMakeLists.txt (hunk 4 above)
[the remark this answers is not in the document]
Comment (dalg24) [reply core/src/CMakeLists.txt:101]: I was able to reproduce with GCC on Unix.

File: core/unit_test/TestPolicyConstruction.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (dalg24) [reply core/unit_test/TestPolicyConstruction.hpp:793]: This can be proposed in a follow up PR.

**Comment by mhoemmen** [comment]
@PhilMiller If you have a benchmark that's easy to build and run, I'd be happy to test on Windows with MSVC.  The current OpenMP performance situation there is not good -- there's some promise that future MSVC versions will improve it, but a `std::thread` back-end would be worth trying there.

**Comment by PhilMiller** [comment]
> @PhilMiller If you have a benchmark that's easy to build and run, I'd be happy to test on Windows with MSVC. The current OpenMP performance situation there is not good -- there's some promise that future MSVC versions will improve it, but a `std::thread` back-end would be worth trying there.

I believe ExaMiniMD is the go-to reference

**Review by fnrizzi — APPROVED** [review]

**Comment by crtrott** [comment]
Retest this please.
